server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,230 @@
+ * and filters out values whose prefix doesn't match the requested key.
+ */
+public class KeyedJsonAtomicFieldData implements AtomicOrdinalsFieldData {
+
+    private final String key;
+    private final AtomicOrdinalsFieldData delegate;
+
+    KeyedJsonAtomicFieldData(String key,
+                             AtomicOrdinalsFieldData delegate) {
+        this.key = key;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        return delegate.ramBytesUsed();
+    }
+
+    @Override
+    public Collection<Accountable> getChildResources() {
+        return delegate.getChildResources();
+    }
+
+    @Override
+    public SortedSetDocValues getOrdinalsValues() {
+        SortedSetDocValues values = delegate.getOrdinalsValues();
+        return KeyedJsonDocValues.create(key, values);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public ScriptDocValues<?> getScriptValues() {
+        return AbstractAtomicOrdinalsFieldData.DEFAULT_SCRIPT_FUNCTION
+            .apply(getOrdinalsValues());
+    }
+
+    @Override
+    public SortedBinaryDocValues getBytesValues() {
+        return FieldData.toString(getOrdinalsValues());
+    }
+
+    static class KeyedJsonDocValues extends AbstractSortedSetDocValues {
+
+        private final BytesRef key;
+        private final SortedSetDocValues delegate;
+
+        /**
+         * The first and last ordinals whose term has 'key' as a prefix.
+         */
+        private final long minOrd;
+        private final long maxOrd;
+
+        public static KeyedJsonDocValues create(String key, SortedSetDocValues delegate) {
+            BytesRef keyBytes = new BytesRef(key);
+
+            long minOrd, maxOrd;
+            try {
+                minOrd = findMinOrd(keyBytes, delegate);
+                maxOrd = minOrd != -1 ? findMaxOrd(keyBytes, delegate) : -1;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+
+            return new KeyedJsonDocValues(keyBytes, delegate, minOrd, maxOrd);
+        }
+
+        private KeyedJsonDocValues(BytesRef key,
+                                   SortedSetDocValues delegate,
+                                   long minOrd,
+                                   long maxOrd) {
+            this.key = key;
+            this.delegate = delegate;
+            this.minOrd = minOrd;
+            this.maxOrd = maxOrd;
+        }
+
+        @Override
+        public long getValueCount() {
+            return delegate.getValueCount();
+        }
+
+        /**
+         * Returns the (un-prefixed) term value for the requested ordinal.
+         *
+         * Note that this method can only be called on ordinals returned from {@link #nextOrd()}.
+         * Otherwise it may attempt to look up values that do not share the correct prefix, which
+         * can result in undefined behavior or an error.
```

Comment:
We should fail with out-of-range ords, some features are going to call this method on out-of-range ords, such as if you pass `min_doc_count: 0` to a `terms` agg. For the record, rebasing ordinals would address this issue (but has some challenges as highlighted in another comment).

Reply:
We should probably document this limitation.

Reply:
This is good to know, I wasn't aware that we just scanned the full ordinals array in some places. I added an explicit check for out-of-range ordinals.

I started to add documentation around the aggregations features that were not supported, but it was hard to justify and came off as unintuitive. I'm going to look into rebasing the global ordinals in a follow-up PR to try to solve the issue (and if that's not possible will brainstorm some compromise/ explanation).